### PR TITLE
Change: Do not copy replication metrics over and over

### PR DESCRIPTION
**Pull request by schreter**
Replication metrics were cloned for each leader metrics update. This is not necessary. Instead, keep a hashmap of replication metrics with *atomic* indices of matching log on individual replicas. Only if anything else changes clone the map.

The change improves the performance of the empty Raft benchmark by about 5% by reducing reallocations of the replication metrics.

Further improvement could be made by reducing the number of atomic operations on the `Arc` backing the `LeaderMetrics`, but that's likely not worth it as of now.

NOTE: This change slightly changes the API of `ReplicationMetrics`. Previously, public access was given to the member `matched`. Now, the member access was replaced with a method `matched()` and a constructor to construct the object. I.e., any user of the `ReplicationMetrics` interested in exact log position on the replica needs to be adjusted to call the method `matched()` instead of just using `matched` member. To prevent such issues in the future, the APIs should not expose members, only accessor functions. This should be taken into consideration in the upcoming API refactoring.

---
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/198)

### CODE REVIEW

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/198)* status: all files reviewed, 1 unresolved discussion (waiting on @ariesdevil, @lichuang, and @schreter)

___
*[openraft/src/replication/mod.rs, line 56 at r1](https://reviewable.io/reviews/datafuselabs/openraft/198#-Mw_JC_t-bQb0p4DTm7m:-Mw_JC_t-bQb0p4DTm7n:b-aihk9v) ([raw file](https://github.com/datafuselabs/openraft/blob/5ba730c96acffad122c33bca916a4d4034b3ef1d/openraft/src/replication/mod.rs#L56)):*
> ```XML
> pub struct ReplicationMetrics {
>     pub(crate) matched_leader_id: Option<LeaderId>,
>     pub(crate) matched_index: AtomicU64,
> ```

Subtle and delicate!

**Review by schreter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/198)* status: 2 of 5 files reviewed, 1 unresolved discussion

___
*[openraft/src/replication/mod.rs, line 56 at r1](https://reviewable.io/reviews/datafuselabs/openraft/198#-Mw_JC_t-bQb0p4DTm7m:-Mw_QBAk7iF50THPJxCI:b-ql53ag) ([raw file](https://github.com/datafuselabs/openraft/blob/5ba730c96acffad122c33bca916a4d4034b3ef1d/openraft/src/replication/mod.rs#L56)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

Subtle and delicate!
</blockquote></details>

Hehe :-). We are using atomics and other low-level stuff in our various sync primitives and data collections extensively to achieve a decent performance (about 1% of the code base). Rust allows us to do that semi-safely - you just need to prove that your unsafe code works, the rest will work.

BTW, you should change your Reviewable default comment disposition from "Blocking" to "Commenting" (just click on the bubble right to the comment which allows you to change disposition and at the bottom is the possibility to change the default disposition).

**Comment by drmingdrmer** [comment]
> BTW, you should change your Reviewable default comment disposition from "Blocking" to "Commenting" (just click on the bubble right to the comment which allows you to change disposition and at the bottom is the possibility to change the default disposition).

🤔  Did not yet find a `bubble` that resembles your description. What's it look like?

**Review by schreter — COMMENTED** [review]
> 🤔  Did not yet find a bubble that resembles your description. What's it look like?

In Reviewable, click on any code line to add a comment. The bubble is at the top right in the input box popup, for you there will be a gray one-way street sign (minus in circle). When you click on it, you can assign a specific disposition for the comment or default at the bottom of the popup.

> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/198#-:-Mw_adhn7sHJjqHIWpx1:b-3gl4ev)* status: 2 of 5 files reviewed, 1 unresolved discussion

**Comment by drmingdrmer** [comment]
I see that! Thank you!

Really hard to notice it... :(

**Comment by schreter** [comment]
> Really hard to notice it... :(

Well, yes. I'd recommend reading (some of) the docs from Reviewable.

It's free for open-source projects, but for our project, we bought enterprise licenses. We use it for all PRs. Now also together with `git spr` (https://github.com/ejoffe/spr) to manage dependent changes (PR on PR) easily (this is a Gerrit-like workflow). The combination of the two tools is already very good, though they are not (yet) integrated. One just has to pay attention not to submit a change in the "wrong" way :-) (dependent PR into its parent PR).

Using `git spr` + Reviewable it's easy to review *and update* (by rebasing) several changes building on top of each other completely independently of each other and still see the lineage of the conversations in the code properly (GitHub discards code comments in case of rebase and in general, comments get hidden if there are too many - extremely hard to follow).

BTW, you can also enable code line number display in Reviewable - just look into "Tips and Tricks" in the documentation chapter.

**Comment by drmingdrmer** [comment]
> It's free for open-source projects, but for our project, we bought enterprise licenses. We use it for all PRs. Now also together with `git spr` (https://github.com/ejoffe/spr) to manage dependent changes (PR on PR) easily (this is a Gerrit-like workflow). The combination of the two tools is already very good, though they are not (yet) integrated. One just has to pay attention not to submit a change in the "wrong" way :-) (dependent PR into its parent PR).

`git spr` sounds interesting:D

And I'll take your advice to have close look at what reviewable can do.

Thank you!
